### PR TITLE
Scaffold react components

### LINKED ISSUE #11: Scaffold SessionManagerField
### Overview
Perquisite work to build out a new CMS component to display session data in the Member admin UI.

See https://github.com/silverstripe/silverstripe-session-manager/issues/12 for the designs and more details.

### Acceptance Criteria

- [x]  Set up build tooling
    - Base configuration on [Module Template](https://github.com/silverstripe/silverstripe-module)
    - Use current release of silverstripe-webpack
- [x]  Create empty SessionManagerField component and Entwine wrapper
- [x]  Create basic SessionManagerField PHP class
- [x]  Create extension to attach SessionManagerField to Member UI
- [x]  Validate that new field is rendering correctly
- [x]  By default this component is viewed by the current user only 

### Notes
- The module template has some out of date dependencies that will need updating as part of this issue. Let’s look to commit the updates back into the module template repository.

### PR DESCRIPTION
**Pull request by bergice**
Resolves https://github.com/silverstripe/silverstripe-session-manager/issues/11

Marked as draft as this is only the scaffolding for the react field. So it might be a good idea to just approve this without merging, then I'll merge the actual rendering of the field back into this branch later before we merge back into `1`.
I'll also update code coverage after this.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+- Maintenance on this module is a shared effort of those who use it
+- To contribute improvements to the code, ensure you raise a pull request and discuss with the module maintainers
+- Please follow the SilverStripe [code contribution guidelines](https://docs.silverstripe.org/en/contributing/code/) and [Module Standard](https://docs.silverstripe.org/en/developer_guides/extending/modules/#module-standard)
+- Supply documentation that follows the [GitHub Flavored Markdown](https://help.github.com/articles/markdown-basics/) conventions
+- When having discussions about this module in issues or pull request please adhere to the [SilverStripe Community Code of Conduct](https://docs.silverstripe.org/en/contributing/code_of_conduct/)
+
+
+## Contributor license agreement
+By supplying code to this module in patches, tickets and pull requests, you agree to assign copyright 
+of that code to MODULE_COPYRIGHT_HOLDER_HERE., on the condition that these code changes are released under the
```
Comment (Cheddam) [inline CONTRIBUTING.md]:
```suggestion
of that code to Silverstripe Ltd., on the condition that these code changes are released under the 
```

File: src/Extensions/MemberExtension.php [hunk 2]
```diff
@@ -41,6 +54,28 @@ public function updateCMSFields(FieldList $fields)
...
+
+        $fields->removeByName(['SessionManagerField']);
+
+        if (!$this->owner->exists() || !$this->currentUserCanViewSessionManagerConfig()) {
+            return $fields;
+        }
+
+        $fields->addFieldToTab(
+            'Root.Main',
+            $methodListField = SessionManagerField::create(
```
Comment (Cheddam) [inline src/Extensions/MemberExtension.php]: I think this is leftover naming from MFA?

```suggestion
            $sessionManagerField = SessionManagerField::create(
```

File: src/Extensions/MemberExtension.php [hunk 3]
```diff
@@ -54,4 +89,60 @@ protected function getSessionLifetime()
 
         return LoginSession::config()->get('default_session_lifetime');
     }
+
+    /**
+     * Determines whether the logged in user has sufficient permission to see the SessionManager config for this Member.
+     *
+     * @return bool
+     */
+    public function currentUserCanViewSessionManagerConfig(): bool
```
Comment (Cheddam) [inline src/Extensions/MemberExtension.php]:
```suggestion
    public function currentUserCanViewSessionsForMember(): bool
```

File: src/Extensions/MemberExtension.php [hunk 4]
```diff
@@ -54,4 +89,60 @@ protected function getSessionLifetime()
...
+    }
+
+    /**
+     * Determines whether the logged in user has sufficient permission
+     * to modify the SessionManager config for this Member.
+     * Note that this is different from being able to _reset_ the config (which administrators can do).
+     *
+     * @return bool
+     */
+    public function currentUserCanEditSessionManagerConfig(): bool
```
Comment (Cheddam) [inline src/Extensions/MemberExtension.php]:
```suggestion
    public function currentUserCanEditSessionsForMember(): bool
```

File: src/Extensions/MemberExtension.php [hunk 5]
```diff
@@ -54,4 +89,60 @@ protected function getSessionLifetime()
...
+        );
+
+        $category = _t(
+            'SilverStripe\\Security\\Permission.PERMISSIONS_CATEGORY',
+            'Roles and access permissions'
+        );
+
+        $description = _t(
+            __CLASS__ . '.SESSION_MANAGER_PERMISSION_DESCRIPTION',
+            'Ability to view and reset registered SessionManager methods for other members.'
```
Comment (Cheddam) [inline src/Extensions/MemberExtension.php]:
```suggestion
            'Ability to view and purge active login sessions for other members.'
```

File: src/Extensions/MemberExtension.php [hunk 6]
```diff
@@ -54,4 +89,60 @@ protected function getSessionLifetime()
...
+    /**
+     * Provides the SessionManager view/reset permission for selection in the permission list in the CMS.
+     *
+     * @return array
+     */
+    public function providePermissions(): array
+    {
+        $label = _t(
+            __CLASS__ . '.SESSION_MANAGER_PERMISSION_LABEL',
+            'View/reset SessionManager configuration for other members'
```
Comment (Cheddam) [inline src/Extensions/MemberExtension.php]:
```suggestion
            'View/purge login sessions for other members'
```

File: src/FormField/SessionManagerField.php [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    /**
+     * Get the registered backup method (if any) from the currently logged in user.
+     *
+     * @return RegisteredMethod|null
+     */
+    protected function getBackupMethod(): ?RegisteredMethod
+    {
+        $backupMethod = MethodRegistry::singleton()->getBackupMethod();
+        return RegisteredMethodManager::singleton()->getFromMember(Security::getCurrentUser(), $backupMethod);
+    }
```
Comment (Cheddam) [inline src/FormField/SessionManagerField.php]: This is definitely left over from MFA 😅 

```suggestion

```

File: client/src/bundles/bundle.scss [hunk 8]
```diff
@@ -0,0 +1,8 @@
+// Import core variables from silverstripe/admin
+@import "variables";
+
+// Import all of your SCSS stylesheets using relative paths from "components"
+// e.g. @import '../components/MyComponent/MyComponent.scss';
```
Comment (Cheddam) [inline client/src/bundles/bundle.scss]: Could scaffold SessionManagerField.scss as part of this PR?

File: src/Extensions/MemberExtension.php [hunk 9]
```diff
@@ -2,17 +2,29 @@
...
+use SilverStripe\Control\Controller;
 use SilverStripe\Control\Session;
 use SilverStripe\Core\Extension;
 use SilverStripe\Forms\FieldList;
 use SilverStripe\Forms\GridField\GridField;
 use SilverStripe\Forms\GridField\GridFieldConfig_Base;
+use SilverStripe\MFA\Authenticator\ChangePasswordHandler;
+use SilverStripe\MFA\Exception\InvalidMethodException;
+use SilverStripe\MFA\FormField\RegisteredMFAMethodListField;
+use SilverStripe\MFA\Model\RegisteredMethod;
```
Comment (Cheddam) [inline src/Extensions/MemberExtension.php]:
```suggestion
```

**Review by Cheddam — CHANGES_REQUESTED** [review]
A few bits to trim and rename, otherwise this is a solid base to build on 👍

File: src/FormField/SessionManagerField.php (hunk 7 above)
Comment (bergice) [reply src/FormField/SessionManagerField.php]: Hehe, oopsie.

**Comment by bergice** [comment]
> A few bits to trim and rename, otherwise this is a solid base to build on 👍

@Cheddam I've fixed up the things you requested. :)

**Review by Cheddam — APPROVED** [review]

**Pull request merged**